sklearn/ensemble/_hist_gradient_boosting/_bitset.pyx

```diff
@@ -0,0 +1,59 @@
+# cython: cdivision=True
+# cython: boundscheck=False
+# cython: wraparound=False
+# cython: language_level=3
+from .common cimport BITSET_INNER_DTYPE_C
+from .common cimport BITSET_DTYPE_C
+from .common cimport X_DTYPE_C
+from .common cimport X_BINNED_DTYPE_C
+
+cdef inline void init_bitset(BITSET_DTYPE_C bitset) nogil: # OUT
```

Comment:
Some explanations what kind of magic is performed here would help my understanding quite a bit. (Though explained magic is no magic anymore:smirk:)

Reply:
What would help you understand here? Maybe you are not familiar with bitsets? https://en.wikipedia.org/wiki/Bit_array

It's one of those things like anonymous functions that are confusing when you don't know they exist, but it's pretty straightforward.

Regarding the multiple functions doing seemingly the same thing: it's basically because of Cython restrictions. I don't like it either, but we should note that this file is definitely going to change / be removed once we extend support for more than 256 categories, so I wouldn't worry about it too much.

Reply:
Any comment or docstring would help. It took me quite some time to wrap my head around it. Looking into the tests helped a bit. Why are we using bitsets here? Why is this data structure optimal?

Reply:
It's optimal to represent sets of integers in [0, n] when n isn't too high. I'll add a  comment with links to the wikipedia entries

Maybe the difficulty of understanding this comes from not reading the code in the "natural" order. It happens to me often when doing reviews of such big PRs. Usually, it all becomes clearer after a second pass over the PR. I think that when onone is familiar with bitsets, and when you start from the beginning (which is actually the splitter in this case), then the comments there are pretty informative and understanding should be smoother

Reply:
(But please LMK if this is still unclear after looking at the splitter ;) )

Reply:
Your new comments help and reading the splitter even more.